### PR TITLE
feat!: call vim notify from logger

### PR DESCRIPTION
**Pull request by eshepelyuk**

### CODE REVIEW

**Comment by eshepelyuk** [comment]
Hello @jose-elias-alvarez 

This is a follow up to #892.

I should highlight very strange issue. Although I ve rebased my PR on top of the latest main branch - after PR creation style and test job failedz but for those files that i had not touch.

So I had to fix them - there is a separate commit for this.
I have no idea why it didnt fail at main branch.

**Comment by eshepelyuk** [comment]
Hello @jose-elias-alvarez 
Since the PR alreade has gone so far , what do you think about replacing log level conviguration from plenary's specific lowe cased strings (i.e. `info`) to neovim constants, i.e. `vim.log.levela.INFO` ?

**Review by jose-elias-alvarez — COMMENTED** [review]
So now that I've tried this, I think there's a few issues, at least with the default `vim.notify` handler:

1. The handler doesn't show the source of the message. A lot of the messages are not 100% clear without the context that they're coming from null-ls, so this is a problem. Are there other plugins using `vim.notify` that have a solution or workaround for this issue?
2. Setting `notify_level` to `"debug"` or `"warn"` makes Neovim hard to use, since each log message creates a `Press ENTER or type command to continue` prompt. This is the case with Plenary when `use_console = "sync"`, but without an async option, it's quite a bit worse.

Assuming we can solve the first issue, here's one idea for the second: we can disable Plenary console logging by default (but keep it available as a potential debug tool), then we can *always* use `vim.notify` for errors and warnings, since these are the most relevant messages. If for whatever reason users want to suppress these messages, they can do so by wrapping `vim.notify` with custom filtering logic. What do you think?

**Comment by jose-elias-alvarez** [comment]
> I should highlight very strange issue. Although I've rebased my PR on top of the latest main branch, after PR creation the style and test jobs failed, but for those files that i had not touch.

I'm seeing this too, and it's because our StyLua action isn't pinned to a specific version, so it automatically upgraded to version 14. I want to upgrade to the new version, but I don't like some of the changes, so I'm trying to figure out the right config options to get the previous behavior back. Once I do, I'll rebase this PR.

**Comment by eshepelyuk** [comment]
Thanks for feedback, unable to address, closing.

**Comment by jose-elias-alvarez** [comment]
Up to you! I think this PR is almost there - we could just add `null-ls` somewhere into the string, which is how [nvim-lspconfig does it](https://github.com/neovim/nvim-lspconfig/blob/7be59acd6e3afdfecf426b5ed6b3d12956157a90/lua/lspconfig.lua#L84). Either way, thanks for the work on this.

**Comment by eshepelyuk** [comment]
> Up to you! I think this PR is almost there - we could just add `null-ls` somewhere into the string, which is how [nvim-lspconfig does it](https://github.com/neovim/nvim-lspconfig/blob/7be59acd6e3afdfecf426b5ed6b3d12956157a90/lua/lspconfig.lua#L84). Either way, thanks for the work on this.



> Up to you! I think this PR is almost there - we could just add `null-ls` somewhere into the string, which is how [nvim-lspconfig does it](https://github.com/neovim/nvim-lspconfig/blob/7be59acd6e3afdfecf426b5ed6b3d12956157a90/lua/lspconfig.lua#L84). Either way, thanks for the work on this.

@jose-elias-alvarez then i am lost :((
i had a feeling that there is a requirement of one more round of development with reviews and more round etc without the clear end.

if i am wrong and you mean smth simple - maybe you could rephrase or just create more descriptive list of tasks to make this complete.

**Comment by jose-elias-alvarez** [comment]
This is pretty much what I had in mind (the diff is based on your PR but compared to `main`):

```diff
diff --git a/lua/null-ls/logger.lua b/lua/null-ls/logger.lua
index 22933be..dbe86cb 100644
--- a/lua/null-ls/logger.lua
+++ b/lua/null-ls/logger.lua
@@ -1,6 +1,14 @@
 local c = require("null-ls.config")
 local u = require("null-ls.utils")
 
+local plenary_to_vim = {
+    trace = vim.log.levels.TRACE,
+    debug = vim.log.levels.DEBUG,
+    info = vim.log.levels.INFO,
+    warn = vim.log.levels.WARN,
+    error = vim.log.levels.ERROR,
+}
+
 local log = {}
 
 --- Adds a log entry using Plenary.log
@@ -11,20 +19,29 @@ function log:add_entry(msg, level)
         return
     end
 
+    level = level:lower()
+
+    local notify_level = plenary_to_vim[level] or -1
+    -- always notify about warnings and errors
+    if notify_level >= vim.diagnostic.severity.WARN then
+        -- add simple title to make sure that users can see the source of the message w/ the default handler
+        -- I wonder if there's a better way to handle this? but I think it's okay for now
+        vim.notify(string.format("[null-ls] %s", msg), notify_level)
+    end
+
+    -- avoid messing with plenary handle, since it uses metatable magic to show a proper trace for log messages
     if self.__handle then
-        -- plenary uses lower-case log levels
-        self.__handle[level:lower()](msg)
+        self.__handle[level](msg)
         return
     end
 
     local default_opts = {
         plugin = "null-ls",
         level = c.get().log.level or "warn",
-        use_console = c.get().log.use_console,
+        use_console = false,
         info_level = 4,
     }
     if c.get().debug then
-        default_opts.use_console = false
         default_opts.level = "trace"
     end
 
@@ -34,7 +51,7 @@ function log:add_entry(msg, level)
     end
 
     local handle = plenary_log.new(default_opts)
-    handle[level:lower()](msg)
+    handle[level](msg)
     self.__handle = handle
 end
```

The basic changes are:
- Disable Plenary's console log completely
- Always notify the user about errors and warnings
- Use a template string for messages passed to `vim.notify` so that users using the default handler can see the message's source

Feel free to reopen this if you'd like to pick it up again, but if not, I can implement the changes myself. It's technically a breaking change, but I think it's a good one.

File: Makefile [hunk 1]
```diff
@@ -1,8 +1,12 @@
 .PHONY: test
 test:
+	mkdir -p ~/.cache/nvim/
```
[the remark this answers is not in the document]
Comment (ievgenii-shepeliuk) [reply Makefile:3]: tests are failing on GitHub unless I manually create `cache` folder.

[the remark this answers is not in the document]
Comment (eshepelyuk) [reply Makefile:3]: Tests are failing at GitHub if cache folder is missing.

File: doc/CONFIG.md [hunk 2]
```diff
@@ -144,22 +142,23 @@ severity. See `:help diagnostic-severity` for available values.
 
 ### log (table)
 
-Sets options for null-ls logging.
+Enables or disables logging altogether. Setting this to `false` or `nil` will suppress
```
[the remark this answers is not in the document]
Comment (eshepelyuk) [reply doc/CONFIG.md]: No valid anymore

**Comment by eshepelyuk** [comment]
@jose-elias-alvarez rebasee on main and updated the logic. please check.

File: Makefile (hunk 1 above)
[the remark this answers is not in the document]
Comment (ievgenii-shepeliuk) [reply Makefile]: Actually fixed this by adjusting test.

File: doc/CONFIG.md [hunk 3]
```diff
@@ -52,11 +52,8 @@ local defaults = {
     default_timeout = 5000,
     diagnostics_format = "#{m}",
     fallback_severity = vim.diagnostic.severity.ERROR,
-    log = {
-        enable = true,
-        level = "warn",
-        use_console = "async",
-    },
+    log_level = "warn",
+    notify_format = "[null-ls] %s",
```
Comment (jose-elias-alvarez) [inline doc/CONFIG.md:56]: I like the idea of making this configurable!

**Review by jose-elias-alvarez — APPROVED** [review]
LGTM! Thanks for your work and patience on this!

**Comment by eshepelyuk** [comment]
Great news,thnx !

**Pull request merged**